File: AEDS/BASES/Ordenacao/shellsort.c
#include <stdio.h>

void insercao (int vet[], int n, int gap, int alt) {
    for (int i = (alt + gap); i < n; i += gap) {
        int aux = vet[i];
        int j = i - alt;
        while (j >= 0 && vet[j] > aux) {
            vet[j + alt] = vet[j];
            j -= alt;
        }
        vet[j + alt] = aux;
    }
}

void shellsort(int vet[], int n) {
    int alt = 1;

    do {
        alt = (alt * 3) + 1;
    } while (alt < n);
    do {
        alt /= 3;
        for(int gap = 0; gap < alt; gap++) insercao(vet, n, gap, alt);
    } while (alt != 1);
}

File: AEDS/BASES/Ordenacao/shellsort.java
public class shellsort {
    public static void insercao (int array[], int n, int gap, int alt) {
        for (int i = (alt + gap); i < n; i += alt) {
            int aux = array[i];
            int j = i - alt;
            while (j >= 0 && array[j] > aux) {
                array[j + alt] = array[j];
                j -= alt;
            }
            array[j + alt] = aux;
        }
    }
    
    public static void shell (int array[], int n) {
        int alt = 1;
        do {
            alt = (alt * 3) + 1;
        } while (alt < n);

        do {
            alt /= 3;
            for (int gap = 0; gap < alt; gap++) insercao(array, n, gap, alt);
        } while (alt > 1);
    }

}
